polylaue/ui/editor.py

```diff
@@ -0,0 +1,339 @@
+# Copyright © 2024, UChicago Argonne, LLC. See "LICENSE" for full details.
+
+import logging
+
+from PySide6.QtWidgets import (
+    QDialog,
+    QDialogButtonBox,
+    QDoubleSpinBox,
+    QFileDialog,
+    QHBoxLayout,
+    QLabel,
+    QLineEdit,
+    QMessageBox,
+    QPushButton,
+    QSpinBox,
+    QStyle,
+    QVBoxLayout,
+    QGridLayout,
+    QWidget,
+)
+
+from polylaue.model.editable import (
+    Editable,
+    ParameterType,
+    ParameterDescription,
+    ValidationError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Field:
+    def __init__(self, description: ParameterDescription):
+        self._description = description
+        self._widget = self.create_widget()
+
+    def create_widget(self) -> QWidget:
+        raise NotImplementedError()
+
+    @property
+    def description(self) -> ParameterDescription:
+        return self._description
+
+    @property
+    def widget(self) -> QWidget:
+        return self._widget
+
+    @property
+    def value(self):
+        raise NotImplementedError()
+
+    @value.setter
+    def value(self, v):
+        raise NotImplementedError()
+
+
+class IntegerField(Field):
+    def create_widget(self) -> QWidget:
+        sb = QSpinBox()
+        minimum = self._description.get("min")
+        maximum = self._description.get("max")
+        if minimum is not None:
+            sb.setMinimum(int(minimum))
+        if maximum is not None:
+            sb.setMaximum(int(maximum))
+
+        self._spin_box = sb
+
+        return sb
+
+    @property
+    def value(self):
+        return self._spin_box.value()
+
+    @value.setter
+    def value(self, v):
+        self._spin_box.setValue(v)
+
+
+class FloatField(Field):
+    def create_widget(self) -> QWidget:
+        sb = QDoubleSpinBox()
```

Comment:
In the dialog, we ought to use `ScientificDoubleSpinBox` (located [here](https://github.com/PolyLaue/PolyLaue/blob/1e03ae9b44a794365dcae1199011574666258ff2/polylaue/ui/scientificspinbox.py#L72)) rather than `QDoubleSpinBox`.

The `QDoubleSpinBox` requires us to fix the number of decimal places, but the `ScientificDoubleSpinBox` does not. `ScientificDoubleSpinBox` also allows floats to be specified in scientific notation, and it automatically eliminates trailing zeros (which `QDoubleSpinBox` does not).